veracruz-client/src/veracruz_client.rs

```diff
@@ -11,26 +11,96 @@
-use rustls::{Certificate, ClientConnection, PrivateKey};
-use rustls_pemfile;
 use std::{
     convert::TryFrom,
     io::{Read, Write},
     path::Path,
-    str::from_utf8,
+    sync::{Arc, Mutex},
 };
 use veracruz_utils::VERACRUZ_RUNTIME_HASH_EXTENSION_ID;
 use webpki;
 
-#[derive(Debug)]
+/// VeracruzClient struct. The remote_session_id is shared between
+/// VeracruzClient and InsecureConnection so that it is available from
+/// VeracruzClient methods and can also be updated by the
+/// InsecureConnection methods invoked by mbedtls.
 pub struct VeracruzClient {
-    tls_connection: ClientConnection,
-    remote_session_id: Option<u32>,
+    tls_connection: mbedtls::ssl::Context<InsecureConnection>,
+    remote_session_id: Arc<Mutex<Option<u32>>>,
     policy: Policy,
     policy_hash: String,
-    package_id: u32,
-    client_cert: String,
+}
+
+/// This is the structure given to mbedtls and used for reading and
+/// writing cyphertext, using the standard Read and Write traits.
+struct InsecureConnection {
+    read_buffer: Vec<u8>,
+    veracruz_server_url: String,
+    remote_session_id: Arc<Mutex<Option<u32>>>,
+}
+
+impl Read for InsecureConnection {
+    fn read(&mut self, data: &mut [u8]) -> Result<usize, std::io::Error> {
+        // Return as much data from the read_buffer as fits.
+        let n = std::cmp::min(data.len(), self.read_buffer.len());
+        data[0..n].clone_from_slice(&self.read_buffer[0..n]);
+        self.read_buffer = self.read_buffer[n..].to_vec();
+        Ok(n)
+    }
+}
+
+impl Write for InsecureConnection {
+    // An error returned from this function gets lost in any case
+    // (perhaps because of the C library) so we just use "unwrap".
+    fn write(&mut self, data: &[u8]) -> Result<usize, std::io::Error> {
+        // Send all the data to the server.
+        let string_data = base64::encode(&data);
+        let combined_string = format!(
+            "{:} {:}",
+            self.remote_session_id.lock().unwrap().unwrap_or(0),
+            string_data
+        );
+        let dest_url = format!("http://{:}/runtime_manager", self.veracruz_server_url,);
+        // Spawn a separate thread so that we can use reqwest::blocking.
+        let body = std::thread::spawn(move || {
+            let client_build = reqwest::blocking::ClientBuilder::new().build().unwrap();
+            let ret = client_build
+                .post(dest_url)
+                .body(combined_string)
+                .send()
+                .map_err(|_| {
+                    std::io::Error::new(std::io::ErrorKind::Other, "reqwest send failed")
+                })?;
+            if ret.status() != reqwest::StatusCode::OK {
+                return Err(std::io::Error::new(
+                    std::io::ErrorKind::Other,
+                    "reqwest bad status",
+                ));
+            }
+            Ok(ret.text().unwrap())
+        })
+        .join()
+        .unwrap()?;
```

Comment:
Similarly here.  We seem to be calling `unwrap` but then also propagating another error with `?`.

Reply:
Ditto. A call to `write` doesn't correspond very naturally to a particular call into the C library, I think.